### PR TITLE
Makefile: reimplement to delegate behavior to `./dev`

### LINKED ISSUE #75734: Makefile: add warnings advising people to use `dev` in place of `make build`, `make test`, etc.
We don't have to cover every single use case for the `Makefile`, but covering the following (and smartly recommending the corresponding `dev` command) would be a good start:

```
make build{,short}
make test
make lint{,short}
... others?
```

Epic CRDB-8036

### PR DESCRIPTION
**Pull request by knz**
Epic: CRDB-8349
Follow up to #84400.
Informs #104804.

This commit acknowledges the now-widespread use of `./dev`
inside the CockroachDB developer community, and switches
the `Makefile` over to delegate its behavior to `./dev`.

(We continue offering a `Makefile` so that newcomers who
reach out to `make` both get an OK behavior, and a hint to use
`./dev`).

Also, the `Makefile` is renamed to `GNUmakefile` since it is not
meant to work with non-GNU make. (This is a no-op for users
of GNU make already, since GNU make reaches for `GNUmakefile`
automatically if it exists)

Release note (build change): The top-level `Makefile` was replaced
by a stub `GNUmakefile` which defers its behavior to `dev`.

The common targets `make [all]`, `make test` and `make install` remain
for compatibility with most unix installation guides.

The previous make rules remain available via `make -C
build/GNUmakefile.obsolete ...`.

### CODE REVIEW

**Comment by knz** [comment]
marking as draft pending removal of the github CI target (as discussed on slack)

File: GNUmakefile [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+help:
+	@echo
+	@echo "Tip: use ./dev instead of 'make'."
+	@echo "Try:"
+	@echo "    ./dev help"
+	@echo
+
+# Generic build rules.
+.PHONY: build build%
+build: doctor
```
Comment (rickystewart) [inline GNUmakefile:27]: Let's not have this direct dependency on `doctor` in this target or any others.

If you run `dev build` without running `doctor`, you'll get an error and it'll ask you to run `doctor`. Meanwhile, `dev build` does a (lightweight) check whether you've already run `doctor` every time you run it.

However, the way this dependency is structured means that we must unconditionally run `dev doctor` every time before building, which is rather slow.

**Review by rickystewart — APPROVED** [review]

File: GNUmakefile (hunk 1 above)
Comment (knz) [reply GNUmakefile]: > Let's not have this direct dependency on doctor in this target or any others.

Fixed.

**Comment by knz** [comment]
TFYR!

bors r=rickystewart

**Pull request merged**